pkg/executor/join/hash_table_v2.go

```diff
@@ -0,0 +1,216 @@
+	prev := st.hashTable[pos]
+	st.hashTable[pos] = rowAddress
+	setNextRowAddress(rowAddress, prev)
+}
+
+func (st *subTable) atomicUpdateHashValue(pos uint64, rowAddress unsafe.Pointer) {
+	for {
+		prev := atomic.LoadPointer(&st.hashTable[pos])
+		if atomic.CompareAndSwapPointer(&st.hashTable[pos], prev, rowAddress) {
+			setNextRowAddress(rowAddress, prev)
+			break
+		}
+	}
+}
+
+func (st *subTable) build(startSegmentIndex int, endSegmentIndex int) {
+	if startSegmentIndex == 0 && endSegmentIndex == len(st.rowData.segments) {
+		for i := startSegmentIndex; i < endSegmentIndex; i++ {
+			for _, index := range st.rowData.segments[i].validJoinKeyPos {
+				rowAddress := st.rowData.segments[i].rowLocations[index]
+				hashValue := st.rowData.segments[i].hashValues[index]
+				pos := hashValue & st.posMask
+				st.updateHashValue(pos, rowAddress)
+			}
+		}
+	} else {
+		for i := startSegmentIndex; i < endSegmentIndex; i++ {
+			for _, index := range st.rowData.segments[i].validJoinKeyPos {
+				rowAddress := st.rowData.segments[i].rowLocations[index]
+				hashValue := st.rowData.segments[i].hashValues[index]
+				pos := hashValue & st.posMask
+				st.atomicUpdateHashValue(pos, rowAddress)
+			}
+		}
+	}
+}
+
+type hashTableV2 struct {
+	tables          []*subTable
+	partitionNumber uint64
+}
+
+type rowPos struct {
+	subTableIndex   int
+	rowSegmentIndex int
+	rowIndex        uint64
+}
+
+type rowIter struct {
+	table      *hashTableV2
+	currentPos *rowPos
+	endPos     *rowPos
+}
+
+func (ri *rowIter) getValue() unsafe.Pointer {
+	return ri.table.tables[ri.currentPos.subTableIndex].rowData.segments[ri.currentPos.rowSegmentIndex].rowLocations[ri.currentPos.rowIndex]
+}
+
+func (ri *rowIter) next() {
+	ri.currentPos.rowIndex++
+	if ri.currentPos.rowIndex == uint64(ri.table.tables[ri.currentPos.subTableIndex].rowData.segments[ri.currentPos.rowSegmentIndex].rowCount()) {
+		ri.currentPos.rowSegmentIndex++
+		ri.currentPos.rowIndex = 0
+		for ri.currentPos.rowSegmentIndex == len(ri.table.tables[ri.currentPos.subTableIndex].rowData.segments) {
+			ri.currentPos.subTableIndex++
+			ri.currentPos.rowSegmentIndex = 0
+			if ri.currentPos.subTableIndex == int(ri.table.partitionNumber) {
+				break
+			}
+		}
+	}
+}
+
+func (ri *rowIter) isEnd() bool {
+	return !(ri.currentPos.subTableIndex < ri.endPos.subTableIndex || ri.currentPos.rowSegmentIndex < ri.endPos.rowSegmentIndex || ri.currentPos.rowIndex < ri.endPos.rowIndex)
+}
+
+func newJoinHashTableForTest(partitionedRowTables []*rowTable) *hashTableV2 {
+	// first make sure there is no nil rowTable
+	jht := &hashTableV2{
+		tables:          make([]*subTable, len(partitionedRowTables)),
+		partitionNumber: uint64(len(partitionedRowTables)),
+	}
+	for i, rowTable := range partitionedRowTables {
+		jht.tables[i] = newSubTable(rowTable)
+	}
+	return jht
+}
+
+func (jht *hashTableV2) createRowPos(pos uint64) *rowPos {
+	if pos < 0 || pos > jht.totalRowCount() {
```

Comment:
pos is impossible to be lower than 0 as it's type is `uint64`

Reply:
remove the check of `pos < 0`